### PR TITLE
Leader election

### PR DESCRIPTION
**Pull request by bkielbasa**
This is a draft of leader election. The algorithm looks like this:

* While starting the server - connect to ZooKeeper and create  Protected, Ephemeral and Sequential znode
* get the node ID
* check if it's the first on the list
* if yes - make it master
* otherwise - it's a slave

Because it's in-memory db I do not care about restarting the cluster (for now)

### CODE REVIEW

**Comment by bkielbasa** [comment]
@mmierzwa :)

File: server/leaderelection.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		}
+	}
+
+	return true, nil
+}
+
+func (le *leaderElection) waitForChildrenChanged(events <-chan zconst.Event) {
+	select {
+	case <- events:
+		children, _, e, _ := le.zkconn.ChildrenW(electionZNode)
```
Comment (mmierzwa) [inline server/leaderelection.go]: Returning 4 values from a single function looks bizarre, but with 3rd party library you cannot do anything with that beside a good naming and separation :)

File: server/leaderelection.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+const rootZNode = "/genki"
+const electionZNode = rootZNode + "/leader_election"
+const election = electionZNode + "/guid_"
+
+const RoleUnknown = 0
+const RoleMaster = 1
+const RoleSlave = 2
+
+type leaderElection struct {
+	zkconn     *zk.Conn
```
[the remark this answers is not in the document]
Comment (bkielbasa) [reply server/leaderelection.go]: yeah, I think it's lack of tests for the `leaderElection`. I'll have to add some. About the name of the struct - I'd prefer to leave it as it is. I try to follow these rules: https://talks.golang.org/2014/names.slide
I'll rethink some variable's names ;)

File: server/leaderelection.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+		return -1, errors.Wrapf(err, "cannot create %s folder on ZooKeeper", electionZNode)
+	}
+
+	s, err := le.zkconn.CreateProtectedEphemeralSequential(election, []byte{}, zk.WorldACL(zk.PermAll))
+	if err != nil {
+		return -1, errors.Wrap(err, "cannot create folder on ZooKeeper")
+	}
+
+	parts := strings.Split(s, "guid_")
+	index, err := strconv.Atoi(parts[1])
```
[the remark this answers is not in the document]
Comment (bkielbasa) [reply server/leaderelection.go]: WDYT about renaming it to nNum for node number?

File: server/leaderelection.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		if i < index {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func (le *leaderElection) waitForChildrenChanged(events <-chan zconst.Event) {
+	select {
```
[the remark this answers is not in the document]
Comment (bkielbasa) [reply server/leaderelection.go]: I think I have to add a timeout here and retry after X secs. Good point!

File: server/leaderelection.go (hunk 1 above)
Comment (bkielbasa) [reply server/leaderelection.go]: yeah, I don't like it too but I did not need those values :(

**Comment by bkielbasa** [comment]
I removed server_test file because I have to rewrite them to for example integrational test.

File: server/leaderelection.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mmierzwa) [reply server/leaderelection.go]: I rather thought about something like `nodeIndex` or `nodeNumber` - prefer full names over abbreviations :)

**Pull request merged**